src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/FlowStateAllowedEventService.java

```diff
@@ -0,0 +1,96 @@
+import static uk.gov.hmcts.reform.unspec.callback.CaseEvent.RESPOND_EXTENSION;
+import static uk.gov.hmcts.reform.unspec.callback.CaseEvent.WITHDRAW_CLAIM;
+import static uk.gov.hmcts.reform.unspec.service.flowstate.MainFlowState.CLAIM_ISSUED;
+import static uk.gov.hmcts.reform.unspec.service.flowstate.MainFlowState.CLAIM_STAYED;
+import static uk.gov.hmcts.reform.unspec.service.flowstate.MainFlowState.DRAFT;
+import static uk.gov.hmcts.reform.unspec.service.flowstate.MainFlowState.EXTENSION_REQUESTED;
+import static uk.gov.hmcts.reform.unspec.service.flowstate.MainFlowState.EXTENSION_RESPONDED;
+import static uk.gov.hmcts.reform.unspec.service.flowstate.MainFlowState.FULL_DEFENCE;
+import static uk.gov.hmcts.reform.unspec.service.flowstate.MainFlowState.RESPONDED_TO_CLAIM;
+import static uk.gov.hmcts.reform.unspec.service.flowstate.MainFlowState.SERVICE_ACKNOWLEDGED;
+import static uk.gov.hmcts.reform.unspec.service.flowstate.MainFlowState.SERVICE_CONFIRMED;
+
+@Service
+@RequiredArgsConstructor
+public class FlowStateAllowedEventService {
+
+    private final StateFlowEngine stateFlowEngine;
+
+    private static final Map<String, List<CaseEvent>> ALLOWED_EVENTS_ON_FLOW_STATE = Map.of(
+        DRAFT.fullName(),
+        List.of(CREATE_CASE, WITHDRAW_CLAIM, DISCONTINUE_CLAIM),
+
+        CLAIM_ISSUED.fullName(),
+        List.of(MOVE_TO_STAYED, CONFIRM_SERVICE, WITHDRAW_CLAIM, DISCONTINUE_CLAIM),
+
+        CLAIM_STAYED.fullName(),
+        List.of(WITHDRAW_CLAIM, DISCONTINUE_CLAIM),
+
+        SERVICE_CONFIRMED.fullName(),
+        List.of(ACKNOWLEDGE_SERVICE, DEFENDANT_RESPONSE, WITHDRAW_CLAIM, DISCONTINUE_CLAIM),
+
+        SERVICE_ACKNOWLEDGED.fullName(),
+        List.of(REQUEST_EXTENSION, DEFENDANT_RESPONSE, WITHDRAW_CLAIM, DISCONTINUE_CLAIM),
+
+        EXTENSION_REQUESTED.fullName(),
+        List.of(DEFENDANT_RESPONSE, RESPOND_EXTENSION, WITHDRAW_CLAIM, DISCONTINUE_CLAIM),
+
+        EXTENSION_RESPONDED.fullName(),
+        List.of(DEFENDANT_RESPONSE, WITHDRAW_CLAIM, DISCONTINUE_CLAIM),
+
+        RESPONDED_TO_CLAIM.fullName(),
+        List.of(CLAIMANT_RESPONSE, WITHDRAW_CLAIM, DISCONTINUE_CLAIM),
+
+        FULL_DEFENCE.fullName(),
+        List.of(WITHDRAW_CLAIM, DISCONTINUE_CLAIM)
+    );
+
+    public MainFlowState getFlowState(CaseData caseData) {
+        StateFlow stateFlow = stateFlowEngine.evaluate(caseData);
+        return MainFlowState.fromFullName(stateFlow.getState().getName());
+    }
+
+    public List<CaseEvent> getAllowedEvents(String stateFullName) {
+        return ALLOWED_EVENTS_ON_FLOW_STATE.getOrDefault(stateFullName, emptyList());
+    }
+
+    public boolean isAllowedOnState(String stateFullName, CaseEvent caseEvent) {
+        return ALLOWED_EVENTS_ON_FLOW_STATE
+            .getOrDefault(stateFullName, emptyList())
+            .contains(caseEvent);
+    }
+
+    public boolean isAllowed(CaseDetails caseDetails, CaseEvent caseEvent) {
```

Comment:
Again as above I would move this to another service.

Reply:
This is using `isAllowedOnState(...)` which is using `ALLOWED_EVENTS_ON_FLOW_STATE` so I think it better belongs here.